### PR TITLE
Added the capability to post rich messages to slack

### PR DESCRIPTION
**Pull request by raddessi**
### What does this PR do?
Adds the optional `blocks` and `attachments` params to the `slack.post_message` module function so messages can be made easier to read.

I'm not sure how to test this as there is not current test against the module directly, only a few for the state. I can add some to there if you tell me what you would like verified in the tests.

### What issues does this PR fix or reference?
Fixes: NA

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [X] Docs
- [X] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
We need tests covering this change please.

**Comment by raddessi** [comment]
Definitely @s0undt3ch can do, but I had asked how to proceed on that since there are no module level tests currently. What would you like verified?

**Comment by s0undt3ch** [comment]
> there are no module level tests currently

😞 

That at least the queries are properly built with the added functionality (empty or with data)

**Comment by raddessi** [comment]
OK sounds good, actually now that you all started the pytest move I'll just try to set up a basic module test file. Thanks!

**Comment by raddessi** [comment]
Ok that's done but I have a few Qs:

1. I created a changelog entry with the PR id since there is no Issue related to this.. is that correct? The docs say to use the issue number only.
2. How do I update docs, or is this even needed in this case? I thought the module docs were pulled directly from the docstrings.

**Comment by raddessi** [comment]
BTW, thank you again for the work switching to nox/pytest. It's a gigantic improvement :1st_place_medal:

**Comment by raddessi** [comment]
> Ok that's done but I have a few questions:
> 
>     1. I created a changelog entry with the PR id since there is no Issue related to this.. is that correct? The docs say to use the issue number only.
> 
>     2. How do I update docs, or is this even needed in this case? I thought the module docs were pulled directly from the docstrings.

Bumping @s0undt3ch

**Comment by s0undt3ch** [comment]
> I created a changelog entry with the PR id since there is no Issue related to this.. is that correct? 

Yes

> The docs say to use the issue number only.
    How do I update docs, or is this even needed in this case? I thought the module docs were pulled directly from the docstrings.

They are. You should update the docs/docstring with what changed, etc.
For example, when where the new options added, which version, etc....

>  I'll just try to set up a basic module test file.

Awesome!

> BTW, thank you again for the work switching to nox/pytest. It's a gigantic improvement 🥇

Thank You!

File: tests/pytests/unit/modules/test_slack.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+"""
+Tests for salt.modules.slack module
+"""
+
+import pytest
+import salt.modules.slack_notify as slack
+from salt.ext.six.moves.urllib.parse import urlencode as _urlencode
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_slack.py:7]:
```suggestion
from urllib.parse import urlencode as _urlencode
```

Maybe even better,
```suggestion
import urllib.parse
```
And use the fully qualified module name and function

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/pytests/unit/modules/test_slack.py (hunk 1 above)
Comment (raddessi) [reply tests/pytests/unit/modules/test_slack.py:7]: Great! Thanks for the info, I'll fix this and get back to you.

Comment (raddessi) [reply tests/pytests/unit/modules/test_slack.py]: Done

**Comment by raddessi** [comment]
@s0undt3ch It looks like that test failure was a flake?

File: salt/modules/slack_notify.py [hunk 2]
```diff
@@ -166,15 +166,22 @@ def find_user(name, api_key=None):
...
 
 
-def post_message(channel, message, from_name, api_key=None, icon=None):
+def post_message(
+    channel, message, from_name, api_key=None, icon=None, attachments=None, blocks=None,
+):
     """
     Send a message to a Slack channel.
 
+    .. versionchanged:: 3002.2
```
Comment (raddessi) [inline salt/modules/slack_notify.py:175]: Actually, what should this version be set to?

**Comment by raddessi** [comment]
Nice! Thanks for retriggering that one

**Comment by raddessi** [comment]
Hey @s0undt3ch and @sagetherage I would love it if this could be included in the next release. Thanks!

File: salt/modules/slack_notify.py (hunk 2 above)
Comment (s0undt3ch) [reply salt/modules/slack_notify.py:175]: 3003

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Comment by raddessi** [comment]
Done! Is there a place you all post the next expected release version that we could reference in cases like this?

**Review by Ch3LL — APPROVED** [review]

**Comment by Ch3LL** [comment]
You can look at teh release notes documentation: https://docs.saltproject.io/en/latest/topics/releases/index.html

As you can see the most recent previous release was 3002.2 so the next feature release would be 3003. Hope that helps :)

**Comment by raddessi** [comment]
> You can look at teh release notes documentation: https://docs.saltproject.io/en/latest/topics/releases/index.html
> 
> As you can see the most recent previous release was 3002.2 so the next feature release would be 3003. Hope that helps :)

Hmm.. I know the last release, but I mean how do I know if the next release will be a point release like `3002.3` or a full release like `3003`? I know the next major version is 3003, but how do I know if this would go in to a patch or a full release? Just assume full release on new features and patch on something being fixed and then wait for acceptance to fully determine it?

**Comment by raddessi** [comment]
Looks like some tests failed to OOM on jenkins?

**Comment by sagetherage** [comment]
ok updating and running tests again

**Comment by s0undt3ch** [comment]
> Just assume full release on new features and patch on something being fixed and then wait for acceptance to fully determine it?

Yup!

**Review by s0undt3ch — APPROVED** [review]

**Review by garethgreenaway — APPROVED** [review]

**Pull request merged**